### PR TITLE
Horovod

### PR DESCRIPTION
**Pull request by rahul003**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: example/image-classification/common/data.py [hunk 1]
```diff
@@ -99,8 +99,8 @@ def __init__(self, num_classes, data_shape, max_iter, dtype):
         self.dtype = dtype
         label = np.random.randint(0, num_classes, [self.batch_size,])
         data = np.random.uniform(-1, 1, data_shape)
-        self.data = mx.nd.array(data, dtype=self.dtype, ctx=mx.Context('cpu_pinned', 0))
-        self.label = mx.nd.array(label, dtype=self.dtype, ctx=mx.Context('cpu_pinned', 0))
+        self.data = mx.nd.array(data, dtype=self.dtype, ctx=mx.Context('cpu', 0))
+        self.label = mx.nd.array(label, dtype=self.dtype, ctx=mx.Context('cpu', 0))
```
Comment (rahul003) [inline example/image-classification/common/data.py:103]: Why?

Comment (ctcyang) [reply example/image-classification/common/data.py:103]: Haibin and I found a bug where cpu_pinned will cause GPU0 to be allocated even though the process requested another GPU, say GPU 7. This ends up using up way too much memory on GPU0 that standard batch sizes won't be able to fit. It's a bug that due to: (a) cpu_pinned not calling cudaSetDevice before allocation, (b) cudaSetDevice(0) being called in src code in more than one place.
